adapters/intertech/intertech.go

```diff
@@ -0,0 +1,244 @@
+package intertech
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v3/adapters"
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+)
+
+const (
+	pageIDMacro = "{{page_id}}"
+	impIDMacro  = "{{imp_id}}"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+type ExtImpIntertech struct {
+	PageID int `json:"page_id"`
+	ImpID  int `json:"imp_id"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var requests []*adapters.RequestData
+
+	referer := getReferer(request)
+	cur := getCur(request)
+
+	for _, imp := range request.Imp {
+		extImp, err := parseAndValidateImpExt(imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		modifiedImp, err := modifyImp(imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		modifiedUrl, err := a.modifyUrl(extImp, referer, cur)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		modRequest := *request
+		modRequest.Imp = []openrtb2.Imp{modifiedImp}
+
+		reqData, err := buildRequestData(modRequest, modifiedUrl)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		requests = append(requests, reqData)
+	}
+
+	return requests, errs
+}
+
+func parseAndValidateImpExt(imp openrtb2.Imp) (ExtImpIntertech, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := jsonutil.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return ExtImpIntertech{}, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp #%s: unable to parse bidder ext: %s", imp.ID, err),
+		}
+	}
+
+	var extImp ExtImpIntertech
+	if err := jsonutil.Unmarshal(bidderExt.Bidder, &extImp); err != nil {
+		return ExtImpIntertech{}, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp #%s: unable to parse intertech ext: %s", imp.ID, err),
+		}
+	}
+
+	return extImp, nil
+}
+
+func modifyImp(imp openrtb2.Imp) (openrtb2.Imp, error) {
+	if imp.Banner != nil {
+		banner, err := updateBanner(imp.Banner)
+		if err != nil {
+			return openrtb2.Imp{}, &errortypes.BadInput{
+				Message: fmt.Sprintf("imp #%s: %s", imp.ID, err.Error()),
+			}
+		}
+		imp.Banner = banner
+		return imp, nil
+	}
+	if imp.Native != nil {
+		return imp, nil
+	}
+	return openrtb2.Imp{}, &errortypes.BadInput{
+		Message: fmt.Sprintf("Intertech only supports banner and native types. Ignoring imp id=%s", imp.ID),
+	}
+}
+
+func updateBanner(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
+	if banner == nil {
+		return nil, fmt.Errorf("banner is null")
+	}
```

Comment:
You can delete this conditional as `banner` will always be not nil since this function call is only made if banner is not nil.